content/4_Gold/Digit_DP.mdx

```diff
@@ -92,61 +115,280 @@ void reset() {
+
+		/*
+		 * If we want to count how many sequences have exactly half digits of
+		 * targ and half of targ2, then any other number than targ or targ2
+		 * doesn't matter
+		 */
 		if (is_started && targ2 != -1 && i != targ && i != targ2) { continue; }
-		ans += solve_dp(loc + 1, k + (is_started ? (targ == i ? 1 : -1) : 0),
-		                is_largest, is_started, targ, targ2);
+
+		/*
+		 * The count for target digit targ is increased by one if i is targ, or
+		 * decreased by one otherwise.
+		 */
+		int new_k = k;
+		if (is_started) {
+			if (targ == i) {
+				new_k = k + 1;
+			} else {
+				new_k = k - 1;
+			}
+		}
+		ans += solve_dp(pos + 1, new_k, is_under, is_started, targ, targ2);
 	}
-	return dp[loc][k][largest][started] = ans;
+	return dp[pos][k][under][started] = ans;
 }
 
-ll count_interesting_to(ll x) {
-	num = to_string(x);
+/** Count interesting sequences that are less or equal ubound */
+ll count_interesting_to(ll ubound) {
+	num = to_string(ubound);
 	ll ans = 0;
 	for (int i = 0; i <= 9; i++) {
 		reset();
-		ans += solve_dp(0, 20, true, false, i, -1);
+		ans += solve_dp(0, 20, false, false, i, -1);
 	}
-	ll dup = 0;
+
+	/*
+	 * If a sequence's digits consist of exactly half i and half j, then it will
+	 * be counted twice. We have to subtract duplicates to avoid overcount.
+	 */
+	ll duplicates = 0;
 	for (int i = 0; i <= 9; i++) {
 		for (int j = 0; j <= 9; j++) {
 			reset();
-			dup += solve_dp(0, 20, true, false, i, j);
+			duplicates += solve_dp(0, 20, false, false, i, j);
 		}
 	}
-	return ans - (dup / 2);
+	return ans - (duplicates / 2);
 }
 
 int main() {
-	ll x, y;
-	cin >> x >> y;
-	cout << count_interesting_to(y) - count_interesting_to(x - 1) << endl;
+	freopen("odometer.in", "r", stdin);
+	freopen("odometer.out", "w", stdout);
+
+	ll X, Y;
+	cin >> X >> Y;
+	cout << count_interesting_to(Y) - count_interesting_to(X - 1) << endl;
 }
 ```
+
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Odometer {
+	// dp[pos][count][under][started]
+	long[][][][] dp = new long[19][50][2][2];
+	String num;
+
+	/** Reset the dp array to its initial values. */
+	public void reset() {
+		for (int i = 0; i < 19; i++) {
+			for (int j = 0; j < 50; j++) {
+				for (int k = 0; k < 2; k++) {
+					for (int l = 0; l < 2; l++) { dp[i][j][k][l] = -1; }
+				}
+			}
+		}
+	}
+
+	/**
+	 * Calculate the number of sequences whose digits consist of at least half
+	 * targ. If targ2 is not -1, then calculate the number of sequences whose
+	 * digits are exactly half targ and half targ2.
+	 *
+	 * @param pos     the starting position
+	 * @param k       counter for the number of the target digit targ
+	 * @param under   1 if the current sequence is smaller than the given
+	 *     upperbound
+	 * @param started 1 if the sequence has any digit other than leading zeros
+	 */
+	public long solveDP(int pos, int k, int under, int started, int targ,
+	                    int targ2) {
```

Comment:
bro what is this function formatting :sob:

Reply:
![image](https://user-images.githubusercontent.com/44814054/236620815-c38d7915-8cb3-4049-b6a8-91cb9253c032.png)
sry ci hates it